panic on Compute

Hi Andrey,

We were using the v2.5.1 for some time but upgraded to v3.0.1 recently.

After the upgrade, we hit the following panic:

```
panic: runtime error: index out of range [1678240084308591512] with length 0

goroutine 46 [running]:
github.com/puzpuzpuz/xsync/v3.(*MapOf[...]).doCompute(0x12d3540, 0x2e, 0xc0006e5f80, 0x0, 0x1)
	/home/runner/go/pkg/mod/github.com/puzpuzpuz/xsync/v3@v3.0.1/mapof.go:267 +0xa59
github.com/puzpuzpuz/xsync/v3.(*MapOf[...]).Compute(...)
	/home/runner/go/pkg/mod/github.com/puzpuzpuz/xsync/v3@v3.0.1/mapof.go:215
```

We haven't seen this panic for quite some time, and it occurred right after the upgrade, so it might be related to the recent changes.

The map has the following type, if it is going to help: `xsync.NewMapOfPresized[int64, someStruct](1000)`, where `someStruct` has two fields, one of the type `any` and one of the type `int`

--- a/map.go
+++ b/map.go
@@ -481,7 +481,7 @@ func (m *Map) resize(knownTable *mapTable, hint mapResizeHint) {
 		newTable = newMapTable(tableLen << 1)
 	case mapShrinkHint:
 		shrinkThreshold := int64((tableLen * entriesPerMapBucket) / mapShrinkFraction)
-		if table.sumSize() <= shrinkThreshold {
+		if tableLen > minMapTableLen && table.sumSize() <= shrinkThreshold {
 			// Shrink the table with factor of 2.
 			atomic.AddInt64(&m.totalShrinks, 1)
 			newTable = newMapTable(tableLen >> 1)

--- a/map_test.go
+++ b/map_test.go
@@ -470,6 +470,37 @@ func TestMapStoreThenLoadAndDelete(t *testing.T) {
 	}
 }
 
+func TestMapStoreThenParallelDelete_DoesNotShrinkBelowMinTableLen(t *testing.T) {
+	const numEntries = 1000
+	m := NewMap()
+	for i := 0; i < numEntries; i++ {
+		m.Store(strconv.Itoa(i), i)
+	}
+
+	cdone := make(chan bool)
+	go func() {
+		for i := 0; i < numEntries; i++ {
+			m.Delete(strconv.Itoa(int(i)))
+		}
+		cdone <- true
+	}()
+	go func() {
+		for i := 0; i < numEntries; i++ {
+			m.Delete(strconv.Itoa(int(i)))
+		}
+		cdone <- true
+	}()
+
+	// Wait for the goroutines to finish.
+	<-cdone
+	<-cdone
+
+	stats := CollectMapStats(m)
+	if stats.RootBuckets < MinMapTableLen {
+		t.Fatalf("table was too small: %d", stats.RootBuckets)
+	}
+}
+
 func sizeBasedOnRange(m *Map) int {
 	size := 0
 	m.Range(func(key string, value interface{}) bool {

--- a/mapof.go
+++ b/mapof.go
@@ -430,7 +430,7 @@ func (m *MapOf[K, V]) resize(knownTable *mapOfTable[K, V], hint mapResizeHint) {
 		newTable = newMapOfTable[K, V](tableLen << 1)
 	case mapShrinkHint:
 		shrinkThreshold := int64((tableLen * entriesPerMapBucket) / mapShrinkFraction)
-		if table.sumSize() <= shrinkThreshold {
+		if tableLen > minMapTableLen && table.sumSize() <= shrinkThreshold {
 			// Shrink the table with factor of 2.
 			atomic.AddInt64(&m.totalShrinks, 1)
 			newTable = newMapOfTable[K, V](tableLen >> 1)

--- a/mapof_test.go
+++ b/mapof_test.go
@@ -515,6 +515,37 @@ func TestMapOfStructStoreThenLoadAndDelete(t *testing.T) {
 	}
 }
 
+func TestMapOfStoreThenParallelDelete_DoesNotShrinkBelowMinTableLen(t *testing.T) {
+	const numEntries = 1000
+	m := NewMapOf[int, int]()
+	for i := 0; i < numEntries; i++ {
+		m.Store(i, i)
+	}
+
+	cdone := make(chan bool)
+	go func() {
+		for i := 0; i < numEntries; i++ {
+			m.Delete(i)
+		}
+		cdone <- true
+	}()
+	go func() {
+		for i := 0; i < numEntries; i++ {
+			m.Delete(i)
+		}
+		cdone <- true
+	}()
+
+	// Wait for the goroutines to finish.
+	<-cdone
+	<-cdone
+
+	stats := CollectMapOfStats(m)
+	if stats.RootBuckets < MinMapTableLen {
+		t.Fatalf("table was too small: %d", stats.RootBuckets)
+	}
+}
+
 func sizeBasedOnTypedRange(m *MapOf[string, int]) int {
 	size := 0
 	m.Range(func(key string, value int) bool {

